lib/xapi-stdext-unix/unixext.ml

```diff
@@ -161,12 +161,17 @@ let string_of_file file_path = Buffer.contents (buffer_of_file file_path)
 let atomic_write_to_file fname perms f =
   let module Filenameext = Xapi_stdext_std.Filenameext in
   let tmp = Filenameext.temp_file_in_dir fname in
-  Unix.chmod tmp perms;
+  let tmp_perms = 0o600 in
+  Unix.chmod tmp tmp_perms;
   finally
     (fun () ->
-       let fd = Unix.openfile tmp [Unix.O_WRONLY; Unix.O_CREAT] perms (* ignored since the file exists *) in
+       (* file permissions in openfile are ignored since the file exists *)
+       let fd = Unix.openfile tmp [Unix.O_WRONLY; Unix.O_CREAT] tmp_perms in
```

Comment:
If the file is expected to exist, can we drop the O_CREAT here?

Reply:
Is it expected to exist? seems like an implementation detail of the function. I would rather have temp_file_in_dir return a file descriptor.

Reply:
Returning an fd would be a better choice indeed. If it doesn't exist yet then it cannot be safely created here though because another thread/process could try to create a file with the same name. To safely create temp files O_EXCL must be used (which `temp_file_in_dir` does), O_EXCL only works when the file doesn't exist, so we cannot use it here.

Reply:
As we are discovering, creation of temp files is tricky. I would prefer to use `Stdlib.Filename.temp_file` (which creates the file and returns its name) because I expect it to be better tested than our own code.

Reply:
It's the only usage of the function in the whole toolstack.
I'm partial to using `Filename.open_temp_file`
We could change xenopsd and xcp-rrd to actually use that new function instead of their own copy.

Reply:
Agreed, `Filename.open_temp_file ~temp_dir` is a better alternative and we can just delete `temp_file_in_dir`.